File: dcp-solutions/2024-10-03/solution.go
```markdown
// Definition for a binary tree node.
type TreeNode struct {
	Val   int
	Left  *TreeNode
	Right *TreeNode
}

func NodeVal(v interface{}) *TreeNode {
	return &TreeNode{Val: v}
}

type LockingTree struct {
	root *TreeNode
}

func (lt *LockingTree) dfs(node *TreeNode, locked map[*TreeNode]bool) bool {
	if node == nil {
		return false
	}
	locked[node] = true

	if !lt.dfs(node.Left, locked) || !lt.dfs(node.Right, locked) {
		return false
	}

	return true
}

func (lt *LockingTree) isLocked(node *TreeNode) bool {
	return node != nil && node.Val != 0
}

func (lt *LockingTree) lock(node *TreeNode) bool {
	if lt.isLocked(node) {
		return false
	}

	if lt.dfs(node, map[*TreeNode]bool{}) == false {
		node.Val = 1 // 0 for unlocked and positive value if locked
		return true
	}

	return false
}

func (lt *LockingTree) unlock(node *TreeNode) bool {
	if lt.isLocked(node) {
		return false
	}

	if node.Val != 0 {
		node.Val = 0
		return true
	}

	return false
}
```

File: dcp-solutions/2024-10-03/solution.rb
```ruby
class TreeNode
  attr_accessor :val, :left, :right, :parent, :locked

  def initialize(val = 0, left = nil, right = nil)
    @val = val
    @left = left
    @right = right
    @parent = nil
    @locked = false
  end

  def is_locked
    return @locked
  end

  def lock
    return false if @locked || (@left && !@left.locked) || (@right && !@right.locked)
    @locked = true
    return true
  end

  def unlock
    return false if not(@locked || (@left && !@left.locked) || (@right && !@right.locked))
    @locked = false
    return true
  end
end
```